README.md
# progress_dialog

A light weight package to show progress dialog. As it is a stateful widget, you can change the text shown on the dialog dynamically.

[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![](https://img.shields.io/badge/dontate-Buy%20Me%20a%20Coffee-blueviolet)](https://www.buymeacoffee.com/fayaz) &nbsp; 
[![LinkedIn](https://img.shields.io/badge/LinkedIn-in-0e76a8)](https://www.linkedin.com/in/fayaz07) &nbsp; [![Follow](https://img.shields.io/github/followers/fayaz07?style=social)](https://github.com/fayaz07) &nbsp; [![Fork](https://img.shields.io/github/forks/fayaz07/progress_dialog?style=social)](https://github.com/fayaz07/progress_dialog/fork) &nbsp; [![Star](https://img.shields.io/github/stars/fayaz07/progress_dialog?style=social)](https://github.com/fayaz07/progress_dialog/star) &nbsp; [![Watches](https://img.shields.io/github/watchers/fayaz07/progress_dialog?style=social)](https://github.com/fayaz07/progress_dialog/) 

[![Get the library](https://img.shields.io/badge/Get%20library-pub-blue)](https://pub.dev/packages/progress_dialog) &nbsp; [![Example](https://img.shields.io/badge/Example-Ex-success)](https://pub.dev/packages/progress_dialog#-example-tab-)



## Supported Dart Versions
**Dart SDK version >= 2.1.0**


## Demo

<img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/progress_dialog_1.png" height="35%" width="35%"  alt="Normal dialog Demo"/> <img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/progress_dialog_2.png"   height="35%" width="35%" alt="Download dialog Demo" />


## Installation
[![Pub](https://img.shields.io/badge/pub-1.2.0-blue)](https://pub.dev/packages/progress_dialog)

Add the Package
```yaml
dependencies:
  progress_dialog: ^1.2.0
```

## How to use



Import the package in your dart file

```dart
import 'package:progress_dialog/progress_dialog.dart';
```
Create an instance of ProgressDialog
```dart
ProgressDialog pr;
```

Initialise the **pr** object inside the **build()** method passing context to it

<ol>
<li> Initialize the ProgressDialog object <br>  

```dart
pr = new ProgressDialog(context);
```
</li>


<li> By default it is a normal dialog to show some message, if you would like to use it to show percentage of progress done, specify the optional <b>type</b> parameter and specify if you want your dialog to dismiss when back button is pressed <b>isDismissible</b> parameter (Optional) <br/>

```dart
//For normal dialog
pr = new ProgressDialog(context,type: ProgressDialogType.Normal, isDismissible: true/false, showLogs: true/false);
    
//For showing progress percentage
pr = new ProgressDialog(context,type: ProgressDialogType.Download, isDismissible: true/false, showLogs: true/false);
```
</li>

  
<li>Style the progress dialog (Optional)<br/>

```dart
pr.style(
  message: 'Downloading file...',
  borderRadius: 10.0,
  backgroundColor: Colors.white,
  progressWidget: CircularProgressIndicator(),
  elevation: 10.0,
  insetAnimCurve: Curves.easeInOut,
  progress: 0.0,
  maxProgress: 100.0,
  progressTextStyle: TextStyle(
     color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400),
  messageTextStyle: TextStyle(
     color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600)
  );
```
```note
Note: You don't need to use all parameters, all of them are optional
```

</li>

<li>Showing the progress dialog<br>
  
```dart
pr.show();
```
</li>

<li>
Dynamically update the content shown out there <br/>
  
```dart
pr.update(
  progress: 50.0,
  message: "Please wait...",
  progressWidget: Container(
    padding: EdgeInsets.all(8.0), child: CircularProgressIndicator()),
  maxProgress: 100.0,
  progressTextStyle: TextStyle(
    color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400),
  messageTextStyle: TextStyle(
    color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600),
  );
```
```note
Note: You don't need to use all parameters, all of them are optional
```
</li>


<li>
Dismissing the progress dialog<br/>
  
```dart
pr.hide().then((isHidden) {
  print(isHidden);
});
```

</li>  

</ol>

> Navigating to next screens must be done after the completion of Future - hide(). [See here for example](https://gist.github.com/fayaz07/52dfe04d9f8594501465f169aa6ef6b6)

### Check if progress dialog is showing

```dart
bool isProgressDialogShowing = pr.isShowing();
print(isProgressDialogShowing);
```

---

## Demo

<img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/pd_demo_nd.gif" alt="Normal dialog Demo"/> <img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/pd_demo_dd.gif" alt="Download dialog Demo" />


### Default configuration/styles

If you don't like to configure/style the dialog and continue with the default style, it's okay but just have a look at our default configuration.

| Attribute           | Value   |
| -------------       | :-----:|
| Dismissible         |    true |
| ProgressDialogType  |    ProgressDialogType.Normal |
| BackgroundColor     | Colors.white |
| BorderRadius        |   RoundedRectangularBorder(radius: 8.0) |
| AnimationCurve      |    Curves.easeInOut |
| Elevation           |    8.0 |
| ProgressWidget      |    Double_rings_loding_indicator |
| MessageTextStyle    |    color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600 |
| ProgressTextStyle   |    color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400 |
| showLogs            |    false |

#### Well let's discuss limits for configuring it

| Attribute           | Can be updated during instantiating   | Can be updated during styling   |Can be updated during dialog is shown |
| -------------       | :-------------------------------:     |:------------------------------: |:-------------------------------:     |
| Dismissible         | Yes                                   |No                               |No                                    |
| ProgressDialogType  | Yes                                   |No                               |No                                    |
| BackgroundColor     | No                                    |Yes                              |No                                    |
| BorderRadius        | No                                    |Yes                              |No                                    |
| AnimationCurve      | No                                    |Yes                              |No                                    |
| Elevation           | No                                    |Yes                              |No                                    |
| ProgressWidget      | No                                    |Yes                              |Yes                                   |
| MessageTextStyle    | No                                    |Yes                              |Yes                                   |
| ProgressTextStyle   | No                                    |Yes                              |Yes                                   |
| ShowLogs            | Yes                                   |No                               |No                                    |

---

### Want to contribute? 
Pull requests and issues are always welcome!

#### How to contribute?
<ol>
  <li> Fork the <a href="https://github.com/fayaz07/progress_dialog">repository</a></li>
  <li> Clone it to your local machine </li>
  <li> Open the project in your favourite editor </li>  
  <li> Open cmd/terminal and run <b>flutter clean</b> and then <b>flutter packages get</b> </li>
  <li> Make the changes </li>
  <li> Create a <b>Pull Request</b> </li>
</ol>

#### View the issues [here](https://github.com/fayaz07/progress_dialog/issues)

#### This library is only tested for Android, iOS contributors are most welcome

---
Loading indicator -> https://loading.io/


CHANGELOG.md
## 1.2.1
* Added **await** for showing the dialog
* Added dialog show time delay
## 1.2.0  

 - Added custom styling for widget in runtime, background color,
   elevation, border-radius, text-styles, anim-curve, max-progress,
   no-drop on backbutton press and fixed old bugs

## 1.1.0+1
* Added interactive double ring loading indicator, updated documentation

## 1.0.8
* Added interactive double ring loading indicator, updated documentation

## 1.0.4 
 * Bug fixes.

## 1.0.3 
* Update Dart Version.

## 1.0.2
* Added shown/hidden accessibility.

## 1.0.1
* Added example and ready for package release.


lib/progress_dialog.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

enum ProgressDialogType { Normal, Download, Media }

String _dialogMessage = "Loading...";
double _progress = 0.0, _maxProgress = 100.0;
double _dialogHeight = 100;
double _dialogWidth = 100;

bool _isShowing = false;
BuildContext _context, _dismissingContext;
ProgressDialogType _progressDialogType;
bool _barrierDismissible = true, _showLogs = false;

TextStyle _progressTextStyle = TextStyle(
        color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400),
    _messageStyle = TextStyle(
        color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600);

double _dialogElevation = 8.0, _borderRadius = 8.0;
Color _backgroundColor = Colors.white;
Curve _insetAnimCurve = Curves.easeInOut;

Widget _progressWidget = Image.asset(
  'assets/double_ring_loading_io.gif',
  package: 'progress_dialog',
);

class ProgressDialog {
  _Body _dialog;

  ProgressDialog(BuildContext context,
      {ProgressDialogType type, bool isDismissible, bool showLogs}) {
    _context = context;
    _progressDialogType = type ?? ProgressDialogType.Normal;
    _barrierDismissible = isDismissible ?? true;
    _showLogs = showLogs ?? false;
  }

  void style(
      {double height,
      double width,
      double progress,
      double maxProgress,
      String message,
      Widget progressWidget,
      Color backgroundColor,
      TextStyle progressTextStyle,
      TextStyle messageTextStyle,
      double elevation,
      double borderRadius,
      Curve insetAnimCurve}) {
    if (_isShowing) return;
    if (_progressDialogType == ProgressDialogType.Download) {
      _progress = progress ?? _progress;
    }
    _dialogHeight = height ?? _dialogHeight;
    _dialogWidth = width ?? _dialogWidth;
    _dialogMessage = message ?? _dialogMessage;
    _maxProgress = maxProgress ?? _maxProgress;
    _progressWidget = progressWidget ?? _progressWidget;
    _backgroundColor = backgroundColor ?? _backgroundColor;
    _messageStyle = messageTextStyle ?? _messageStyle;
    _progressTextStyle = progressTextStyle ?? _progressTextStyle;
    _dialogElevation = elevation ?? _dialogElevation;
    _borderRadius = borderRadius ?? _borderRadius;
    _insetAnimCurve = insetAnimCurve ?? _insetAnimCurve;
  }

  void update(
      {double progress,
      double maxProgress,
      String message,
      Widget progressWidget,
      TextStyle progressTextStyle,
      TextStyle messageTextStyle}) {
    if (_progressDialogType == ProgressDialogType.Download) {
      _progress = progress ?? _progress;
    }

    _dialogMessage = message ?? _dialogMessage;
    _maxProgress = maxProgress ?? _maxProgress;
    _progressWidget = progressWidget ?? _progressWidget;
    _messageStyle = messageTextStyle ?? _messageStyle;
    _progressTextStyle = progressTextStyle ?? _progressTextStyle;

    if (_isShowing) _dialog.update();
  }

  bool isShowing() {
    return _isShowing;
  }

  void dismiss() {
    if (_isShowing) {
      try {
        _isShowing = false;
        if (Navigator.of(_dismissingContext).canPop()) {
          Navigator.of(_dismissingContext).pop();
          if (_showLogs) debugPrint('ProgressDialog dismissed');
        } else {
          if (_showLogs) debugPrint('Cant pop ProgressDialog');
        }
      } catch (_) {}
    } else {
      if (_showLogs) debugPrint('ProgressDialog already dismissed');
    }
  }

  Future<bool> hide() {
    if (_isShowing) {
      try {
        _isShowing = false;
        Navigator.of(_dismissingContext).pop(true);
        if (_showLogs) debugPrint('ProgressDialog dismissed');
        return Future.value(true);
      } catch (_) {
        return Future.value(false);
      }
    } else {
      if (_showLogs) debugPrint('ProgressDialog already dismissed');
      return Future.value(false);
    }
  }

  void show() {
    if (!_isShowing) {
      _dialog = new _Body();
      _isShowing = true;

      if (_showLogs) debugPrint('ProgressDialog shown');

      showDialog<dynamic>(
        context: _context,
        barrierDismissible: false,
        builder: (BuildContext context) {
          _dismissingContext = context;
          return WillPopScope(
            onWillPop: () {
              return Future.value(_barrierDismissible);
            },
            child: Dialog(
                backgroundColor: _backgroundColor,
                insetAnimationCurve: _insetAnimCurve,
                insetAnimationDuration: Duration(milliseconds: 100),
                elevation: _dialogElevation,
                shape: RoundedRectangleBorder(
                    borderRadius:
                        BorderRadius.all(Radius.circular(_borderRadius))),
                child: _dialog),
          );
        },
      );
    } else {
      if (_showLogs) debugPrint("ProgressDialog already shown/showing");
    }
  }
}

// ignore: must_be_immutable
class _Body extends StatefulWidget {
  _BodyState _dialog = _BodyState();

  update() {
    _dialog.update();
  }

  @override
  State<StatefulWidget> createState() {
    return _dialog;
  }
}

class _BodyState extends State<_Body> {
  update() {
    setState(() {});
  }

  @override
  void dispose() {
    _isShowing = false;
    if (_showLogs) debugPrint('ProgressDialog dismissed by back button');
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: _dialogHeight,
      width: _dialogWidth,
      child: Row(children: <Widget>[
        Center(
          child: _progressWidget,
        ),
        Visibility(
          visible: _progressDialogType == ProgressDialogType.Media && _dialogMessage != null,
          child: Center(child: Text(_dialogMessage, style: _messageStyle)),
        ),
        Visibility(
          visible: _progressDialogType != ProgressDialogType.Media,
          child: Column(
            children: <Widget>[
              const SizedBox(width: 15.0),
              Expanded(
                child: _progressDialogType == ProgressDialogType.Normal
                    ? Text(_dialogMessage,
                        textAlign: TextAlign.justify, style: _messageStyle)
                    : Stack(
                        children: <Widget>[
                          Positioned(
                            child: Text(_dialogMessage, style: _messageStyle),
                            top: 30.0,
                          ),
                          Positioned(
                            child: Text("$_progress/$_maxProgress",
                                style: _progressTextStyle),
                            bottom: 10.0,
                            right: 10.0,
                          ),
                        ],
                      ),
              ),
            ],
          ),
        ),
      ]),
    );
  }
}
